src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs

```diff
@@ -180,16 +172,34 @@ internal static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationPro
                 protocolSize += protocol.Protocol.Length + 1;
             }
 
-            byte[] buffer = new byte[protocolSize];
+            return protocolSize;
+        }
+
+        internal static void SerializeAlpnProtocolList(List<SslApplicationProtocol> applicationProtocols, Span<byte> buffer)
+        {
+            Debug.Assert(GetAlpnProtocolListSerializedLength(applicationProtocols) == buffer.Length,
+                "GetAlpnProtocolListSerializedSize(applicationProtocols) == buffer.Length");
+
             var offset = 0;
             foreach (SslApplicationProtocol protocol in applicationProtocols)
             {
-                buffer[offset++] = (byte)(protocol.Protocol.Length);
-                protocol.Protocol.Span.CopyTo(buffer.AsSpan(offset));
+                buffer[offset++] = (byte)protocol.Protocol.Length;
+                protocol.Protocol.Span.CopyTo(buffer.Slice(offset));
                 offset += protocol.Protocol.Length;
             }
+        }
 
-            return buffer;
+        internal static unsafe int SslSetAlpnProtos(SafeSslHandle ssl, List<SslApplicationProtocol> applicationProtocols)
+        {
+            int length = GetAlpnProtocolListSerializedLength(applicationProtocols);
+            Span<byte> buffer = length <= 256 ? stackalloc byte[length] : new byte[length];
+            SerializeAlpnProtocolList(applicationProtocols, buffer);
+            return SslSetAlpnProtos(ssl, buffer);
+        }
+
+        internal static unsafe int SslSetAlpnProtos(SafeSslHandle ssl, Span<byte> serializedProtocols)
+        {
+            return SslSetAlpnProtos(ssl, (IntPtr)Unsafe.AsPointer(ref MemoryMarshal.GetReference(serializedProtocols)), serializedProtocols.Length);
```

Comment:
If `serializedProtocols` is a `byte[]`, using `AsPointer` is not safe as the GC could still move it. The buffer should be `fixed` first.